hasprefix.go

```diff
@@ -0,0 +1,40 @@
+package vfs
+
+import (
+	"os"
+	"path/filepath"
+)
+
+// A Stater implements Stat.
+type Stater interface {
+	Stat(string) (os.FileInfo, error)
+}
+
+// HasPrefix returns true if p or any parent of p is the same file as prefix.
+// prefix must exist, but p may not. It is an expensive but accurate alternative
+// to the deprecated filepath.HasPrefix.
+func HasPrefix(fs Stater, p, prefix string) (bool, error) {
+	prefixFI, err := fs.Stat(prefix)
+	if err != nil {
+		return false, err
+	}
+	for {
+		fi, err := fs.Stat(p)
+		switch {
+		case err == nil:
+			if os.SameFile(fi, prefixFI) {
+				return true, nil
+			}
+		case os.IsNotExist(err):
+			// Do nothing and skip ahead to trying p's parent directory.
+		default:
+			return false, err
```

Comment:
IMO this is not correct. If the error is another error (say, a permission denied error), `p` can still be in a subdirectory of `prefix`.

Reply:
Good point again. I've update the code to handle more error cases.